File: Programmers/소수 찾기.cpp
// level 2 
#include <string>
#include <vector>
#include <set>
#include <cmath>

using namespace std;

set<int> per;
int visited[10] = {0, };

int is_prime(int number) {
    if (number == 1 || number == 0)
        return 0;
        
    for (int i = 2; i<=(int)sqrt(number); ++i) {
        if (number % i == 0)
            return 0;
    }
    return 1;
}

void DFS(string numbers, string p, int level, int size) {
    if (level == size) {
        int prime = stoi(p);
        if (is_prime(prime))
            per.insert(prime);
        return;
    }
    
    for (int i = 0; i < numbers.length(); ++i) {
        if (visited[i])
            continue;
        visited[i] = 1;
        p.push_back(numbers[i]);
        DFS(numbers, p, level + 1, size);
        p.pop_back();
        visited[i] = 0;
    }
}

int solution(string numbers) {
    int answer = 0;
    int size = numbers.length();
    
    for (int i = 1; i <= size; ++i) {
        DFS(numbers, "", 0, i);
    }   

    answer = per.size();
    return answer;
}

File: Programmers/소수 찾기.py
from itertools import permutations
import math

# 소수 판별
def is_prime(p):
    if p == 0 or p == 1:
        return False
    
    for i in range(2, int(math.sqrt(p)) + 1):
        if p % i == 0:
            return False
    return True

def solution(numbers):
    answer = 0
    candi = [] # 후보군
    
    # 1부터 길이까지 나올 수 있는 순열
    for i in range(1, len(numbers) + 1):
        per_str = set(permutations(numbers, i))
        # set형과 튜플로 나오니 순서대로 정수를 만들어줌
        for per in per_str:
            tmp = ""
            for pe in per:
                tmp += pe
            candi.append(int(tmp))
    
    for ca in set(candi):            
        if is_prime(ca): # 판별
            answer += 1
    return answer